### PR TITLE
Adding list of devices in jwt. Validating that a user has access to a…

### PR DESCRIPTION
**Pull request by kpears201**
… device for dawg-show. Sending jwt as a cookie to rackProxy/cats server for IR and Power

### CODE REVIEW

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDeviceAccessValidator.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * Checks if a user has access to devices based on the jwt that is in the cookie or authorization header.
+     * If the user is an admin, they have access to all devices.
+     * @param req The request used to get the jwt cookie
+     * @param resp The response used to set a new cookie with the devices
+     * @param refreshDeviceList true if this should go off to find what devices the user has reserved and reset the devices in the cookie
+     * @param deviceIds The list of deviceIds to check
+     * @throws DeviceNotAuthorizedException
+     */
+    public void validateUserHasAccessToDevices(HttpServletRequest req, HttpServletResponse resp, boolean refreshDeviceList, String... deviceIds) throws DeviceNotAuthorizedException {
```
Comment (dimaj) [inline libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDeviceAccessValidator.java:38]: do we really care about validating an array of device ids? If I'm in the `multi` view, wouldn't it make sense to grant access to all devices except for those that are not authorized?

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     * @param req The request object, for getting the cookie
+     * @param resp The response object, for setting the cookie
+     * @return
+     */
+    public DawgJwt assignDevicesToCookie(HttpServletRequest req, HttpServletResponse resp) {
+        String jwt = cookieUtils.extractJwt(req);
+        DawgJwt dawgJwt = jwtEncoder.decodeJwt(jwt);
+        modifyJwtWithDeviceIds(dawgJwt);
+        String jwtWithDevs = jwtEncoder.createUserJWT(dawgJwt);
+        resp.addCookie(cookieUtils.createCookie(jwtWithDevs, -1, req));
```
Comment (dimaj) [inline libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java:40]: should we check for and, if found, delete existing cookie before adding a new one?

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        return dawgJwt;
+    }
+    
+    /**
+     * Gets the deviceIds that are reserved by the logged in user and then puts them in the jwt.
+     * @param dawgJwt The jwt to modify
+     */
+    public void modifyJwtWithDeviceIds(DawgJwt dawgJwt) {
+        String user = dawgJwt.getCreds().getUserName();
```
Comment (dimaj) [inline libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java:49]: should `reserver` be user or some org that a user belongs to?
I am thinking from perspective that if I have multiple devices allocated to me, I wouldn't want to give out my login / password to everyone in my org to get to those devices. However, if devices are reserved by a department (or whatever), and I am part of that department, I can use my own creds and still have access to those devices.

File: libraries/dawg-show/src/main/java/com/comcast/video/dawg/show/key/KeyController.java [hunk 4]
```diff
@@ -44,17 +51,26 @@
...
+    
+    private DawgCookieUtils cookieUtils = new DawgCookieUtils();
+
     @RequestMapping(method = { RequestMethod.POST }, value = "/sendKey")
     @ResponseBody
-    public void sendKey(@RequestParam(value="deviceIds[]") String[] deviceIds, @RequestParam String key, @RequestParam(required = false) String remoteType) throws KeyException, InterruptedException {
-        sendKey(deviceIds, key, null, remoteType);
+    public void sendKey(HttpServletRequest req, HttpServletResponse resp,
+            @RequestParam(value="deviceIds[]") String[] deviceIds, @RequestParam String key, @RequestParam(required = false) String remoteType,
+            @CookieValue(name="dawt", required=false) String dawt) throws KeyException, InterruptedException {
```
Comment (dimaj) [inline libraries/dawg-show/src/main/java/com/comcast/video/dawg/show/key/KeyController.java]: i think `@CookieValue` should be referencing `DawgCookieUtils.COOKIE_NAME` instead

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDeviceAccessValidator.java (hunk 1 above)
Comment (kpears201) [reply libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDeviceAccessValidator.java:38]: I think I would rather just send them to an unauthorized page than have them see a subset of devices

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java (hunk 2 above)
Comment (kpears201) [reply libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java:40]: response.addCookie removes the old cookie with the same name/domain/path

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java (hunk 3 above)
Comment (kpears201) [reply libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java:49]: you are talking about device pooling, it is a feature we have talked about for awhile but will not be solved here.

File: libraries/dawg-show/src/main/java/com/comcast/video/dawg/show/key/KeyController.java (hunk 4 above)
Comment (kpears201) [reply libraries/dawg-show/src/main/java/com/comcast/video/dawg/show/key/KeyController.java]: done

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDeviceAccessValidator.java (hunk 1 above)
Comment (dimaj) [reply libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDeviceAccessValidator.java:38]: ok

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java (hunk 2 above)
Comment (dimaj) [reply libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java:40]: are you sure? I've skimmed through some source code and it wasn't the case... maybe i'm looking at the wrong implementation... what class is the `HttpServletResponse`?

File: libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java (hunk 3 above)
Comment (dimaj) [reply libraries/dawg-server-common/src/main/java/com/comcast/video/dawg/common/security/jwt/JwtDevicePopulator.java:49]: yes, that's what I was talking about.
OK

**Review by dimaj — APPROVED** [review]

**Pull request merged**